readme.md
GS1 barcode generator
=====================

The library generates GS1 barcode. Currently supported formats: GS1 128.

Install
-------

Note that library requires both GD and Free Type extensions installed

```
composer require ayeo/gs1_barcode:1.0.4
```

Usage
-----

The simplest example:

```php
use Ayeo\Barcode;

$builder = new Barcode\Builder();
$builder->output('(10)123456(400)11');
```
This will generate png image using black and white and default font

![barcode](http://q.i-systems.pl/file/fa869375.png "Generated barcode")


Custom barcode:

```php
use Ayeo\Barcode;

$builder = new Barcode\Builder();
$builder->setBarcodeType('gs1-128');
$builder->setFilename('barcode.png');
$builder->setImageFormat('png');
$builder->setWidth(500);
$builder->setHeight(150);
$builder->setFontPath('FreeSans.ttf');
$builder->setFontSize(15);
$builder->setBackgroundColor(255, 255, 255);
$builder->setPaintColor(0, 0, 0);
```

Finally, you can use the output method to stream the image directly to the web browser.

```php
$builder->output('(10)123456(400)11');
```

If you want to save the image file, you can use the saveImage method instead.

```php
$builder->saveImage('(10)123456(400)11');
```

Both output() and saveImage() methods has additional boolean parameters to determine if label should be included on 
the print. Default value is true.

```php
$builder->output('(10)123456(400)11', $withLabel = false);
$builder->saveImage('(10)123456(400)11', $withLabel = false);
```

Actually generate the same barcode because all params all set to default values. This shows only
available settings

Fluent interface is welcome

```php
use Ayeo\Barcode;

Barcode\Builder::build()->setWidth(600)->setBackgroundColor(100, 100, 100)->output('(10)123456(400)11');
```

Additional info
---------------

- Supported image formats: png, jpg
- Barcode must be valid GS1 barcode

Contributing
------------

Everyone is welcome, feel free to join. There is Dockerfile included for ease of developemnt. The image consist of
php5.6 with GD and FreeType extensions. It also contains composer and xdebug. To build the image locally use

```
docker build -t php-gs1 .
```

Supported identifiers
---------------------

The goal is to support all existing gs1 application identifiers but at the moment I have added
only those I needed.
Feel free to add or request some. Full list is [here](http://www.databar-barcode.info/application-identifiers)

|Code       |Name                                        |Min length |Max length |
|-----------|--------------------------------------------|-----------|-----------|
|00         |SERIAL SHIPPING CONTAINER CODE              |18         |18         |
|01         |GLOBAL TRADE ITEM NUMBER                    |14         |14         |
|02         |ITEM TRADE ITEM NUMBER                      |14         |14         |
|10         |BATCH NUMBER                                |1          |20         |
|12         |PAYMENT DATE  (YYMMDD)                      |6          |6          |
|15         |BEST BEFORE DATE (YYMMDD)                   |6          |6          |
|17         |EXPIRATION DATE (YYMMDD)                    |6          |6          |
|37         |NUMBER OF UNITS CONTAINED                   |1          |8          |
|3301       |CONTAINER GROSS WEIGHT (KG)                 |6          |6          |
|390(n)     |AMOUNT PAYABLE - SINGLE MONETARY AREA       |1          |15         |
|400        |CUSTOMER PURCHASE ORDER NUMBER              |1          |30         |
|415        |GLOBAL LOCATION NUMBER OF THE INVOICE PARTY |13         |13         |
|8020       |PAYMENT SLIP REFERENCE NUMBER               |1          |25         |
|96         |COMPANY INTERNAL INFORMATION                |1          |30         |


changelog.md
# 1.0.4 - 2022.04.07

### New

- Functionality of disabling label added


# 1.0.3 - 2020.09.14

### Fixes

- Selecting subset strategy fixed 


# 1.0.2 - 2020.02.26

### Fixes

- Free Type extension checking fixed 


# 1.0.1 - 2018.01.28

### New

- Save to file feature
- Get image as base64
- GS1 codes added:
	- 3900-3909
	- 415
	- 8020
	- 96
	
### Improvements

- System requirements added to composer (gd/imagettftext)
- Composer install info added to readme	


# 1.0.0 - 2015.10.02

Initial version

src/SectionBuilder.php
<?php
namespace Ayeo\Barcode;

use Ayeo\Barcode\Model\Section;

class SectionBuilder
{
    public function build($identifier, $value)
    {
        if (array_key_exists($identifier, $this->data) === false) {
            throw new \LogicException(sprintf('Unknown application identifier %s', $identifier));
        }

        //fixme: the logic below belongs to Section domain

        $sectionData = $this->data[$identifier];

        if (strlen($value) < $sectionData[0]) {
            throw new \LogicException($sectionData[2]);
        }

        if (strlen($value) > $sectionData[1]) {
            throw new \LogicException($sectionData[2]);
        }

        if ($sectionData[0] === $sectionData[1]) {
            $fixedLength = true;
        } else {
            $fixedLength = false;
        }

        return new Section($identifier, $value, $fixedLength);
    }

    private $data = [
        '00' => [18, 18, 'SERIAL SHIPPING CONTAINER CODE (00) - must contains exact 18 digits'],
        '01' => [14, 14, 'GLOBAL TRADE ITEM NUMBER (01) - must contains excat 14 digits'],
        '02' => [14, 14, 'ITEM TRADE ITEM NUMBER (02) - must contains excat 14 digits'],
        '10' => [1, 20, 'BATCH NUMBER (10) - must contains between 1-20 digits'],
        '11' => [6, 6, 'PRODUCTION DATE (11) - must contains exact 6 digits'],
        '12' => [6, 6, 'PAYMENT DATE (12) - must be in YYMMDD format'],
        '15' => [6, 6, 'BEST BEFORE DATE (YYMMDD) - must be in YYMMDD format'],
        '17' => [6, 6, 'EXPIRATION DATE (YYMMDD) - must be in YYMMDD format'],
        '21' => [1, 20, 'SERIAL NUMBER - must containx between 1-20 chars'],
        '37' => [1, 8, 'NUMBER OF UNITS CONTAINED (37) - must contains between 1-8 digits'],
        '240' => [1, 15, 'ADDITIONAL PRODUCT IDENTIFICATION ASSIGNED BY THE MANUFACTURER (240) - must contains between 1-15 digits'
        ],
        '3301' => [6, 6, 'CONTAINER GROSS WEIGHT (KG) - must contains 6 chars'],
		'3900' => [1, 15, 'AMOUNT PAYABLE - SINGLE MONETARY AREA (3900) - must contains between 1-15 digits'],
		'3901' => [1, 15, 'AMOUNT PAYABLE - SINGLE MONETARY AREA (3901) - must contains between 1-15 digits'],
        '3902' => [1, 15, 'AMOUNT PAYABLE - SINGLE MONETARY AREA (3902) - must contains between 1-15 digits'],
		'3903' => [1, 15, 'AMOUNT PAYABLE - SINGLE MONETARY AREA (3903) - must contains between 1-15 digits'],
		'3904' => [1, 15, 'AMOUNT PAYABLE - SINGLE MONETARY AREA (3904) - must contains between 1-15 digits'],
		'3905' => [1, 15, 'AMOUNT PAYABLE - SINGLE MONETARY AREA (3905) - must contains between 1-15 digits'],
		'3906' => [1, 15, 'AMOUNT PAYABLE - SINGLE MONETARY AREA (3906) - must contains between 1-15 digits'],
		'3907' => [1, 15, 'AMOUNT PAYABLE - SINGLE MONETARY AREA (3907) - must contains between 1-15 digits'],
		'3908' => [1, 15, 'AMOUNT PAYABLE - SINGLE MONETARY AREA (3908) - must contains between 1-15 digits'],
		'3909' => [1, 15, 'AMOUNT PAYABLE - SINGLE MONETARY AREA (3909) - must contains between 1-15 digits'],
        '400' => [1, 30, 'CUSTOMER PURCHASE ORDER NUMBER (400) - must contains between 1-30 digits'],
        '403' => [1, 30, 'ROUTE (403) - must contains between 1-30 digits'],
        '415' => [13, 13, 'GLOBAL LOCATION NUMBER OF THE INVOICE PARTY (415) - must contains excat 13 digits'],
        '8020' => [1, 25, 'PAYMENT SLIP REFERENCE NUMBER (8020) - must contains between 1-25 digits'],
        '96' => [1, 30, 'COMPANY INTERNAL INFORMATION (96) - must contains between 1-30 digits'],
    ];
}
